File: tests/artifact_trust/test_artifact_trust.py
from __future__ import annotations

import dataclasses
import json
import unittest
from pathlib import Path

from myactuator_lib.artifact_trust import (
    ZERO_DIGEST,
    ArtifactCandidate,
    ArtifactDecision,
    ArtifactKind,
    ArtifactPolicy,
    ArtifactTrustEngine,
    AuditCommitReceipt,
    DurableCommitReceipt,
    KeyPurpose,
    PersistentState,
    RebootSnapshot,
    VerificationAlgorithm,
    VerificationAssertion,
    empty_persistent_state,
)


ROOT = Path(__file__).resolve().parents[2]
CORPUS = ROOT / "tests/artifact_trust/golden_artifact_trust.jsonl"
D1, D2, D3, D4, D5, D6, D7, D8, D9, DA, DB, DC = (
    character * 64 for character in "123456789abc"
)


def policy(**changes):
    value = ArtifactPolicy(
        platform_profile_selected=True,
        trust_anchor_present=True,
        verifier_bound=True,
        persistent_store_bound=True,
        durable_audit_bound=True,
        expected_kind=ArtifactKind.CONFIGURATION,
        expected_purpose=KeyPurpose.CONFIG_RELEASE,
        expected_algorithm=VerificationAlgorithm.ED25519,
        expected_key_id_digest=D5,
        expected_target_digest=D3,
        expected_envelope_schema_digest=D4,
        minimum_security_epoch=3,
    )
    return dataclasses.replace(value, **changes)


def candidate(**changes):
    value = ArtifactCandidate(
        transaction_digest=D1,
        artifact_digest=D2,
        target_digest=D3,
        envelope_schema_digest=D4,
        key_id_digest=D5,
        kind=ArtifactKind.CONFIGURATION,
        purpose=KeyPurpose.CONFIG_RELEASE,
        algorithm=VerificationAlgorithm.ED25519,
        deployment_sequence=11,
        security_epoch=3,
    )
    return dataclasses.replace(value, **changes)


def assertion(**changes):
    value = VerificationAssertion(
        assertion_digest=D6,
        adapter_id_digest=D7,
        artifact_digest=D2,
        target_digest=D3,
        envelope_schema_digest=D4,
        key_id_digest=D5,
        kind=ArtifactKind.CONFIGURATION,
        purpose=KeyPurpose.CONFIG_RELEASE,
        algorithm=VerificationAlgorithm.ED25519,
        signature_valid=True,
        chain_valid=True,
        key_revoked=False,
    )
    return dataclasses.replace(value, **changes)


def active(**changes):
    value = PersistentState(
        available=True,
        integrity_verified=True,
        active_artifact_digest=D9,
        committed_transaction_digest=DA,
        target_digest=D3,
        envelope_schema_digest=D4,
        key_id_digest=D5,
        kind=ArtifactKind.CONFIGURATION,
        purpose=KeyPurpose.CONFIG_RELEASE,
        algorithm=VerificationAlgorithm.ED25519,
        generation=5,
        deployment_sequence=10,
        security_epoch=3,
    )
    return dataclasses.replace(value, **changes)


def durable(**changes):
    value = DurableCommitReceipt(
        transaction_digest=D1,
        artifact_digest=D2,
        previous_generation=5,
        next_generation=6,
        deployment_sequence=11,
        security_epoch=3,
        write_completed=True,
        readback_verified=True,
    )
    return dataclasses.replace(value, **changes)


def audit(**changes):
    value = AuditCommitReceipt(
        transaction_digest=D1,
        artifact_digest=D2,
        committed_generation=6,
        audit_event_digest=DB,
        durable=True,
    )
    return dataclasses.replace(value, **changes)


def snapshot(**changes):
    value = RebootSnapshot(
        state=active(),
        durable_commit_verified=True,
        audit_commit_verified=True,
    )
    return dataclasses.replace(value, **changes)


def run_variant(variant: str) -> ArtifactDecision:
    selected_policy = policy()
    selected_candidate = candidate()
    selected_assertion: VerificationAssertion | None = assertion()
    selected_active = active()
    selected_durable = durable()
    selected_audit = audit()
    operation = "stage"
    pre_stage = False
    selected_snapshot = snapshot()

    if variant == "pass_stage":
        pass
    elif variant == "profile_not_selected":
        selected_policy = policy(platform_profile_selected=False)
    elif variant == "anchor_missing":
        selected_policy = policy(trust_anchor_present=False)
    elif variant == "verifier_missing":
        selected_policy = policy(verifier_bound=False)
    elif variant == "assertion_missing":
        selected_assertion = None
    elif variant == "signature_invalid":
        selected_assertion = assertion(signature_valid=False)
    elif variant == "chain_invalid":
        selected_assertion = assertion(chain_valid=False)
    elif variant == "key_revoked":
        selected_assertion = assertion(key_revoked=True)
    elif variant == "algorithm_mismatch":
        selected_assertion = assertion(
            algorithm=VerificationAlgorithm.ECDSA_P256_SHA256
        )
    elif variant == "key_id_mismatch":
        selected_assertion = assertion(key_id_digest=DC)
    elif variant == "purpose_mismatch":
        selected_assertion = assertion(
            purpose=KeyPurpose.CALIBRATION_RELEASE
        )
    elif variant == "kind_mismatch":
        selected_assertion = assertion(kind=ArtifactKind.CALIBRATION)
    elif variant == "target_mismatch":
        selected_assertion = assertion(target_digest=DC)
    elif variant == "digest_mismatch":
        selected_assertion = assertion(artifact_digest=DC)
    elif variant == "envelope_mismatch":
        selected_assertion = assertion(envelope_schema_digest=DC)
    elif variant == "persistent_unavailable":
        selected_active = active(available=False)
    elif variant == "persistent_untrusted":
        selected_active = active(integrity_verified=False)
    elif variant == "security_epoch_rollback":
        selected_candidate = candidate(security_epoch=2)
    elif variant == "deployment_sequence_rollback":
        selected_candidate = candidate(deployment_sequence=9)
    elif variant == "duplicate_version_conflict":
        selected_candidate = candidate(deployment_sequence=10)
    elif variant == "invalid_request":
        selected_candidate = candidate(transaction_digest=ZERO_DIGEST)
    elif variant == "stage_occupied":
        pre_stage = True
        selected_candidate = candidate(
            transaction_digest=DC, deployment_sequence=12
        )
    elif variant == "pass_commit":
        operation, pre_stage = "commit", True
    elif variant == "commit_without_stage":
        operation = "commit"
    elif variant == "commit_transaction_mismatch":
        operation, pre_stage = "commit_mismatch", True
    elif variant == "durable_write_incomplete":
        operation, pre_stage = "commit", True
        selected_durable = durable(write_completed=False)
    elif variant == "durable_readback_missing":
        operation, pre_stage = "commit", True
        selected_durable = durable(readback_verified=False)
    elif variant == "durable_transaction_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(transaction_digest=DC)
    elif variant == "durable_artifact_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(artifact_digest=DC)
    elif variant == "durable_sequence_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(deployment_sequence=12)
    elif variant == "durable_epoch_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(security_epoch=4)
    elif variant == "previous_generation_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(previous_generation=4)
    elif variant == "next_generation_mismatch":
        operation, pre_stage = "commit", True
        selected_durable = durable(next_generation=7)
    elif variant == "audit_adapter_missing":
        operation, pre_stage = "commit", True
        selected_policy = policy(durable_audit_bound=False)
    elif variant == "audit_not_durable":
        operation, pre_stage = "commit", True
        selected_audit = audit(durable=False)
    elif variant == "audit_transaction_mismatch":
        operation, pre_stage = "commit", True
        selected_audit = audit(transaction_digest=DC)
    elif variant == "audit_artifact_mismatch":
        operation, pre_stage = "commit", True
        selected_audit = audit(artifact_digest=DC)
    elif variant == "audit_generation_mismatch":
        operation, pre_stage = "commit", True
        selected_audit = audit(committed_generation=7)
    elif variant == "audit_digest_missing":
        operation, pre_stage = "commit", True
        selected_audit = audit(audit_event_digest=ZERO_DIGEST)
    elif variant == "pass_abort":
        operation, pre_stage = "abort", True
    elif variant == "abort_without_stage":
        operation = "abort"
    elif variant == "abort_transaction_mismatch":
        operation, pre_stage = "abort_mismatch", True
    elif variant == "pass_restore":
        operation = "restore"
    elif variant == "restore_integrity_missing":
        operation = "restore"
        selected_snapshot = snapshot(
            state=active(integrity_verified=False)
        )
    elif variant == "restore_durable_missing":
        operation = "restore"
        selected_snapshot = snapshot(durable_commit_verified=False)
    elif variant == "restore_audit_missing":
        operation = "restore"
        selected_snapshot = snapshot(audit_commit_verified=False)
    elif variant == "restore_epoch_rollback":
        operation = "restore"
        selected_snapshot = snapshot(state=active(security_epoch=2))
    elif variant == "restore_digest_missing":
        operation = "restore"
        selected_snapshot = snapshot(
            state=active(active_artifact_digest=ZERO_DIGEST)
        )
    else:
        raise AssertionError(f"unknown corpus variant: {variant}")

    engine = ArtifactTrustEngine(selected_policy, selected_active)
    if pre_stage:
        result = engine.stage(candidate(), assertion())
        assert result.code == ArtifactDecision.PASS_STAGED
    if operation == "stage":
        result = engine.stage(selected_candidate, selected_assertion)
    elif operation == "commit":
        result = engine.commit(D1, selected_durable, selected_audit)
    elif operation == "commit_mismatch":
        result = engine.commit(DC, selected_durable, selected_audit)
    elif operation == "abort":
        result = engine.abort(D1)
    elif operation == "abort_mismatch":
        result = engine.abort(DC)
    elif operation == "restore":
        result = engine.restore(selected_snapshot)
    else:
        raise AssertionError(operation)
    assert result.motion_authorized is False
    return result.code


class ArtifactTrustTests(unittest.TestCase):
    def records(self):
        return [
            json.loads(line)
            for line in CORPUS.read_text().splitlines()
            if line
        ]

    def test_shared_corpus_has_exact_unique_python_results(self):
        records = self.records()
        self.assertEqual(48, len(records))
        self.assertEqual(
            len(records), len({row["variant"] for row in records})
        )
        for record in records:
            with self.subTest(record["variant"]):
                self.assertEqual(
                    ArtifactDecision[record["expected_code"]],
                    run_variant(record["variant"]),
                )

    def test_stage_never_changes_active_and_commit_failures_are_retryable(self):
        engine = ArtifactTrustEngine(policy(), active())
        before = engine.active_state
        staged = engine.stage(candidate(), assertion())
        self.assertEqual(ArtifactDecision.PASS_STAGED, staged.code)
        self.assertEqual(before, engine.active_state)
        self.assertFalse(staged.active_changed)
        failed = engine.commit(
            D1, durable(readback_verified=False), audit()
        )
        self.assertEqual(
            ArtifactDecision.DURABLE_RECEIPT_INVALID, failed.code
        )
        self.assertEqual(before, engine.active_state)
        self.assertEqual(candidate(), engine.staged_candidate)
        committed = engine.commit(D1, durable(), audit())
        self.assertEqual(ArtifactDecision.PASS_COMMITTED, committed.code)
        self.assertTrue(committed.active_changed)
        self.assertIsNone(engine.staged_candidate)
        self.assertEqual(6, engine.active_state.generation)
        self.assertEqual(D2, engine.active_state.active_artifact_digest)
        self.assertEqual(D1, engine.active_state.committed_transaction_digest)

    def test_abort_preserves_active_and_requires_exact_transaction(self):
        engine = ArtifactTrustEngine(policy(), active())
        before = engine.active_state
        engine.stage(candidate(), assertion())
        denied = engine.abort(DC)
        self.assertEqual(ArtifactDecision.TRANSACTION_MISMATCH, denied.code)
        self.assertIsNotNone(engine.staged_candidate)
        passed = engine.abort(D1)
        self.assertEqual(ArtifactDecision.PASS_ABORTED, passed.code)
        self.assertEqual(before, engine.active_state)
        self.assertIsNone(engine.staged_candidate)

    def test_reboot_reconstructs_only_verified_committed_snapshot(self):
        engine = ArtifactTrustEngine(
            policy(), empty_persistent_state()
        )
        denied = engine.restore(
            snapshot(state=active(integrity_verified=False))
        )
        self.assertEqual(
            ArtifactDecision.REBOOT_SNAPSHOT_INVALID, denied.code
        )
        self.assertFalse(engine.active_state.integrity_verified)
        self.assertEqual(0, engine.active_state.generation)
        passed = engine.restore(snapshot())
        self.assertEqual(ArtifactDecision.PASS_RESTORED, passed.code)
        self.assertEqual(active(), engine.active_state)
        denied = engine.restore(snapshot(state=active(target_digest=DC)))
        self.assertEqual(
            ArtifactDecision.REBOOT_SNAPSHOT_INVALID, denied.code
        )

    def test_all_artifact_kinds_require_matching_distinct_purpose(self):
        combinations = (
            (ArtifactKind.CONFIGURATION, KeyPurpose.CONFIG_RELEASE),
            (ArtifactKind.CALIBRATION, KeyPurpose.CALIBRATION_RELEASE),
            (ArtifactKind.FIRMWARE, KeyPurpose.FIRMWARE_RELEASE),
            (ArtifactKind.GENERATED_EVIDENCE, KeyPurpose.EVIDENCE_RELEASE),
        )
        for kind, purpose in combinations:
            selected_policy = policy(
                expected_kind=kind, expected_purpose=purpose
            )
            selected_candidate = candidate(kind=kind, purpose=purpose)
            selected_assertion = assertion(kind=kind, purpose=purpose)
            engine = ArtifactTrustEngine(
                selected_policy, empty_persistent_state()
            )
            self.assertEqual(
                ArtifactDecision.PASS_STAGED,
                engine.stage(selected_candidate, selected_assertion).code,
            )

    def test_core_exposes_assertions_not_crypto_or_private_key_bytes(self):
        fields = {
            field.name
            for structure in (
                ArtifactPolicy,
                ArtifactCandidate,
                VerificationAssertion,
                DurableCommitReceipt,
                AuditCommitReceipt,
            )
            for field in dataclasses.fields(structure)
        }
        self.assertFalse(
            fields
            & {
                "private_key",
                "secret",
                "signature_bytes",
                "public_key_bytes",
                "credential",
            }
        )
        source = (
            ROOT / "host/myactuator_lib/artifact_trust.py"
        ).read_text()
        for forbidden in (
            "from cryptography",
            "import cryptography",
            "import nacl",
            "import Crypto",
        ):
            self.assertNotIn(forbidden, source)


if __name__ == "__main__":
    unittest.main()

File: tests/artifact_trust/test_artifact_trust.cpp
#include <stdio.h>
#include <string.h>

#include "artifact_trust_core.h"

namespace trust = myactuator::artifact_trust;

namespace {

int failures = 0;
#define CHECK(value)                                                        \
    do {                                                                    \
        if (!(value)) {                                                     \
            fprintf(stderr, "CHECK failed at %s:%d: %s\n", __FILE__,      \
                    __LINE__, #value);                                      \
            ++failures;                                                     \
        }                                                                   \
    } while (0)

trust::TrustDigest Digest(uint8_t value) {
    trust::TrustDigest output = {};
    for (size_t index = 0U; index < trust::kTrustDigestSize; ++index)
        output.bytes[index] = value;
    return output;
}

trust::ArtifactPolicy Policy() {
    trust::ArtifactPolicy value = {};
    value.platform_profile_selected = true;
    value.trust_anchor_present = true;
    value.verifier_bound = true;
    value.persistent_store_bound = true;
    value.durable_audit_bound = true;
    value.expected_kind = trust::ArtifactKind::CONFIGURATION;
    value.expected_purpose = trust::KeyPurpose::CONFIG_RELEASE;
    value.expected_algorithm = trust::VerificationAlgorithm::ED25519;
    value.expected_key_id_digest = Digest(5U);
    value.expected_target_digest = Digest(3U);
    value.expected_envelope_schema_digest = Digest(4U);
    value.minimum_security_epoch = 3U;
    return value;
}

trust::ArtifactCandidate Candidate() {
    trust::ArtifactCandidate value = {};
    value.transaction_digest = Digest(1U);
    value.artifact_digest = Digest(2U);
    value.target_digest = Digest(3U);
    value.envelope_schema_digest = Digest(4U);
    value.key_id_digest = Digest(5U);
    value.kind = trust::ArtifactKind::CONFIGURATION;
    value.purpose = trust::KeyPurpose::CONFIG_RELEASE;
    value.algorithm = trust::VerificationAlgorithm::ED25519;
    value.deployment_sequence = 11U;
    value.security_epoch = 3U;
    return value;
}

trust::VerificationAssertion Assertion() {
    trust::VerificationAssertion value = {};
    value.assertion_digest = Digest(6U);
    value.adapter_id_digest = Digest(7U);
    value.artifact_digest = Digest(2U);
    value.target_digest = Digest(3U);
    value.envelope_schema_digest = Digest(4U);
    value.key_id_digest = Digest(5U);
    value.kind = trust::ArtifactKind::CONFIGURATION;
    value.purpose = trust::KeyPurpose::CONFIG_RELEASE;
    value.algorithm = trust::VerificationAlgorithm::ED25519;
    value.signature_valid = true;
    value.chain_valid = true;
    value.key_revoked = false;
    return value;
}

trust::PersistentState Active() {
    trust::PersistentState value = {};
    value.available = true;
    value.integrity_verified = true;
    value.active_artifact_digest = Digest(9U);
    value.committed_transaction_digest = Digest(10U);
    value.target_digest = Digest(3U);
    value.envelope_schema_digest = Digest(4U);
    value.key_id_digest = Digest(5U);
    value.kind = trust::ArtifactKind::CONFIGURATION;
    value.purpose = trust::KeyPurpose::CONFIG_RELEASE;
    value.algorithm = trust::VerificationAlgorithm::ED25519;
    value.generation = 5U;
    value.deployment_sequence = 10U;
    value.security_epoch = 3U;
    return value;
}

trust::DurableCommitReceipt Durable() {
    trust::DurableCommitReceipt value = {};
    value.transaction_digest = Digest(1U);
    value.artifact_digest = Digest(2U);
    value.previous_generation = 5U;
    value.next_generation = 6U;
    value.deployment_sequence = 11U;
    value.security_epoch = 3U;
    value.write_completed = true;
    value.readback_verified = true;
    return value;
}

trust::AuditCommitReceipt Audit() {
    trust::AuditCommitReceipt value = {};
    value.transaction_digest = Digest(1U);
    value.artifact_digest = Digest(2U);
    value.committed_generation = 6U;
    value.audit_event_digest = Digest(11U);
    value.durable = true;
    return value;
}

trust::RebootSnapshot Snapshot() {
    trust::RebootSnapshot value = {};
    value.state = Active();
    value.durable_commit_verified = true;
    value.audit_commit_verified = true;
    return value;
}

enum class Operation {
    STAGE,
    COMMIT,
    COMMIT_MISMATCH,
    ABORT,
    ABORT_MISMATCH,
    RESTORE,
};

trust::ArtifactDecision RunVariant(const char* variant) {
    trust::ArtifactPolicy policy = Policy();
    trust::ArtifactCandidate candidate = Candidate();
    trust::VerificationAssertion assertion = Assertion();
    const trust::VerificationAssertion* assertion_pointer = &assertion;
    trust::PersistentState active = Active();
    trust::DurableCommitReceipt durable = Durable();
    trust::AuditCommitReceipt audit = Audit();
    trust::RebootSnapshot snapshot = Snapshot();
    Operation operation = Operation::STAGE;
    bool pre_stage = false;

    if (strcmp(variant, "pass_stage") == 0) {
    } else if (strcmp(variant, "profile_not_selected") == 0) {
        policy.platform_profile_selected = false;
    } else if (strcmp(variant, "anchor_missing") == 0) {
        policy.trust_anchor_present = false;
    } else if (strcmp(variant, "verifier_missing") == 0) {
        policy.verifier_bound = false;
    } else if (strcmp(variant, "assertion_missing") == 0) {
        assertion_pointer = NULL;
    } else if (strcmp(variant, "signature_invalid") == 0) {
        assertion.signature_valid = false;
    } else if (strcmp(variant, "chain_invalid") == 0) {
        assertion.chain_valid = false;
    } else if (strcmp(variant, "key_revoked") == 0) {
        assertion.key_revoked = true;
    } else if (strcmp(variant, "algorithm_mismatch") == 0) {
        assertion.algorithm =
            trust::VerificationAlgorithm::ECDSA_P256_SHA256;
    } else if (strcmp(variant, "key_id_mismatch") == 0) {
        assertion.key_id_digest = Digest(12U);
    } else if (strcmp(variant, "purpose_mismatch") == 0) {
        assertion.purpose = trust::KeyPurpose::CALIBRATION_RELEASE;
    } else if (strcmp(variant, "kind_mismatch") == 0) {
        assertion.kind = trust::ArtifactKind::CALIBRATION;
    } else if (strcmp(variant, "target_mismatch") == 0) {
        assertion.target_digest = Digest(12U);
    } else if (strcmp(variant, "digest_mismatch") == 0) {
        assertion.artifact_digest = Digest(12U);
    } else if (strcmp(variant, "envelope_mismatch") == 0) {
        assertion.envelope_schema_digest = Digest(12U);
    } else if (strcmp(variant, "persistent_unavailable") == 0) {
        active.available = false;
    } else if (strcmp(variant, "persistent_untrusted") == 0) {
        active.integrity_verified = false;
    } else if (strcmp(variant, "security_epoch_rollback") == 0) {
        candidate.security_epoch = 2U;
    } else if (strcmp(variant, "deployment_sequence_rollback") == 0) {
        candidate.deployment_sequence = 9U;
    } else if (strcmp(variant, "duplicate_version_conflict") == 0) {
        candidate.deployment_sequence = 10U;
    } else if (strcmp(variant, "invalid_request") == 0) {
        candidate.transaction_digest = Digest(0U);
    } else if (strcmp(variant, "stage_occupied") == 0) {
        pre_stage = true;
        candidate.transaction_digest = Digest(12U);
        candidate.deployment_sequence = 12U;
    } else if (strcmp(variant, "pass_commit") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
    } else if (strcmp(variant, "commit_without_stage") == 0) {
        operation = Operation::COMMIT;
    } else if (strcmp(variant, "commit_transaction_mismatch") == 0) {
        operation = Operation::COMMIT_MISMATCH;
        pre_stage = true;
    } else if (strcmp(variant, "durable_write_incomplete") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.write_completed = false;
    } else if (strcmp(variant, "durable_readback_missing") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.readback_verified = false;
    } else if (strcmp(variant, "durable_transaction_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.transaction_digest = Digest(12U);
    } else if (strcmp(variant, "durable_artifact_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.artifact_digest = Digest(12U);
    } else if (strcmp(variant, "durable_sequence_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.deployment_sequence = 12U;
    } else if (strcmp(variant, "durable_epoch_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.security_epoch = 4U;
    } else if (strcmp(variant, "previous_generation_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.previous_generation = 4U;
    } else if (strcmp(variant, "next_generation_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        durable.next_generation = 7U;
    } else if (strcmp(variant, "audit_adapter_missing") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        policy.durable_audit_bound = false;
    } else if (strcmp(variant, "audit_not_durable") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        audit.durable = false;
    } else if (strcmp(variant, "audit_transaction_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        audit.transaction_digest = Digest(12U);
    } else if (strcmp(variant, "audit_artifact_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        audit.artifact_digest = Digest(12U);
    } else if (strcmp(variant, "audit_generation_mismatch") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        audit.committed_generation = 7U;
    } else if (strcmp(variant, "audit_digest_missing") == 0) {
        operation = Operation::COMMIT;
        pre_stage = true;
        audit.audit_event_digest = Digest(0U);
    } else if (strcmp(variant, "pass_abort") == 0) {
        operation = Operation::ABORT;
        pre_stage = true;
    } else if (strcmp(variant, "abort_without_stage") == 0) {
        operation = Operation::ABORT;
    } else if (strcmp(variant, "abort_transaction_mismatch") == 0) {
        operation = Operation::ABORT_MISMATCH;
        pre_stage = true;
    } else if (strcmp(variant, "pass_restore") == 0) {
        operation = Operation::RESTORE;
    } else if (strcmp(variant, "restore_integrity_missing") == 0) {
        operation = Operation::RESTORE;
        snapshot.state.integrity_verified = false;
    } else if (strcmp(variant, "restore_durable_missing") == 0) {
        operation = Operation::RESTORE;
        snapshot.durable_commit_verified = false;
    } else if (strcmp(variant, "restore_audit_missing") == 0) {
        operation = Operation::RESTORE;
        snapshot.audit_commit_verified = false;
    } else if (strcmp(variant, "restore_epoch_rollback") == 0) {
        operation = Operation::RESTORE;
        snapshot.state.security_epoch = 2U;
    } else if (strcmp(variant, "restore_digest_missing") == 0) {
        operation = Operation::RESTORE;
        snapshot.state.active_artifact_digest = Digest(0U);
    } else {
        CHECK(false);
    }

    trust::ArtifactTrustEngine engine(policy, active);
    if (pre_stage) {
        trust::VerificationAssertion pre_assertion = Assertion();
        const trust::ArtifactResult staged =
            engine.stage(Candidate(), &pre_assertion);
        CHECK(staged.code == trust::ArtifactDecision::PASS_STAGED);
    }
    trust::ArtifactResult result = {};
    switch (operation) {
        case Operation::STAGE:
            result = engine.stage(candidate, assertion_pointer);
            break;
        case Operation::COMMIT:
            result = engine.commit(Digest(1U), durable, audit);
            break;
        case Operation::COMMIT_MISMATCH:
            result = engine.commit(Digest(12U), durable, audit);
            break;
        case Operation::ABORT:
            result = engine.abort(Digest(1U));
            break;
        case Operation::ABORT_MISMATCH:
            result = engine.abort(Digest(12U));
            break;
        case Operation::RESTORE:
            result = engine.restore(snapshot);
            break;
    }
    CHECK(!result.motion_authorized);
    const bool pass =
        result.code == trust::ArtifactDecision::PASS_STAGED ||
        result.code == trust::ArtifactDecision::PASS_COMMITTED ||
        result.code == trust::ArtifactDecision::PASS_ABORTED ||
        result.code == trust::ArtifactDecision::PASS_RESTORED;
    CHECK(result.proceed_to_next_gate == pass);
    return result.code;
}

bool Extract(const char* line, const char* prefix, char* output,
             size_t output_size) {
    const char* start = strstr(line, prefix);
    if (start == NULL) return false;
    start += strlen(prefix);
    const char* end = strchr(start, '"');
    if (end == NULL) return false;
    const size_t length = static_cast<size_t>(end - start);
    if (length == 0U || length >= output_size) return false;
    memcpy(output, start, length);
    output[length] = '\0';
    return true;
}

void TestSharedCorpus() {
    FILE* file =
        fopen("tests/artifact_trust/golden_artifact_trust.jsonl", "rb");
    CHECK(file != NULL);
    if (file == NULL) return;
    char line[256] = {};
    char variant[96] = {};
    char expected[96] = {};
    size_t count = 0U;
    while (fgets(line, sizeof(line), file) != NULL) {
        CHECK(strchr(line, '\n') != NULL);
        CHECK(Extract(line, "\"variant\":\"", variant, sizeof(variant)));
        CHECK(Extract(line, "\"expected_code\":\"", expected,
                      sizeof(expected)));
        const trust::ArtifactDecision actual = RunVariant(variant);
        CHECK(strcmp(trust::ArtifactDecisionName(actual), expected) == 0);
        ++count;
    }
    CHECK(!ferror(file));
    fclose(file);
    CHECK(count == 48U);
}

void TestTransactionPreservation() {
    trust::ArtifactPolicy policy = Policy();
    trust::PersistentState state = Active();
    trust::ArtifactTrustEngine engine(policy, state);
    trust::ArtifactCandidate candidate = Candidate();
    trust::VerificationAssertion assertion = Assertion();
    CHECK(engine.stage(candidate, &assertion).code ==
          trust::ArtifactDecision::PASS_STAGED);
    CHECK(engine.activeState().generation == 5U);
    trust::DurableCommitReceipt durable = Durable();
    durable.readback_verified = false;
    trust::AuditCommitReceipt audit = Audit();
    CHECK(engine.commit(Digest(1U), durable, audit).code ==
          trust::ArtifactDecision::DURABLE_RECEIPT_INVALID);
    CHECK(engine.stagePresent());
    CHECK(engine.activeState().generation == 5U);
    durable.readback_verified = true;
    const trust::ArtifactResult committed =
        engine.commit(Digest(1U), durable, audit);
    CHECK(committed.code == trust::ArtifactDecision::PASS_COMMITTED);
    CHECK(committed.active_changed);
    CHECK(!engine.stagePresent());
    CHECK(engine.activeState().generation == 6U);
    CHECK(trust::TrustDigestEqual(engine.activeState().active_artifact_digest,
                                  Digest(2U)));
}

void TestRestoreFailsDisabled() {
    trust::ArtifactPolicy policy = Policy();
    trust::ArtifactTrustEngine engine(
        policy, trust::EmptyPersistentState(true, true));
    trust::RebootSnapshot snapshot = Snapshot();
    snapshot.state.integrity_verified = false;
    CHECK(engine.restore(snapshot).code ==
          trust::ArtifactDecision::REBOOT_SNAPSHOT_INVALID);
    CHECK(engine.activeState().generation == 0U);
    CHECK(!engine.activeState().integrity_verified);
    snapshot = Snapshot();
    CHECK(engine.restore(snapshot).code ==
          trust::ArtifactDecision::PASS_RESTORED);
    CHECK(engine.activeState().generation == 5U);
    snapshot = Snapshot();
    snapshot.state.target_digest = Digest(12U);
    CHECK(engine.restore(snapshot).code ==
          trust::ArtifactDecision::REBOOT_SNAPSHOT_INVALID);
}

}  // namespace

int main() {
    TestSharedCorpus();
    TestTransactionPreservation();
    TestRestoreFailsDisabled();
    CHECK(sizeof(trust::ArtifactTrustEngine) <= 640U);
    CHECK(sizeof(trust::ArtifactResult) <= 32U);
    if (failures != 0) {
        fprintf(stderr, "artifact trust failures=%d\n", failures);
        return 1;
    }
    printf("ARTIFACT_TRUST_CPP_OK corpus=48 engine_bytes=%zu motion=false\n",
           sizeof(trust::ArtifactTrustEngine));
    return 0;
}
